File: client/flutter_app/lib/features/chat/display_effects/display_effects.dart
/// 文本展示效果独立模块（前端）。
///
/// 与服务端 `display-effect-router.ts` 一一对应：服务端按「工具信号 +
/// 内容信号」纯程序路由出 cardType，本模块按 cardType 分发到对应效果组件，
/// 不需要 LLM 参与，也不依赖服务端下发渲染指令。
///
/// 分发入口：[displayEffectsCard]。
/// 各效果组件：
///   - steps      数字步骤链（steps_effect_card.dart）
///   - metric     数据面板网格（metric_effect_card.dart）
///   - carousel   图片轮播横滑（carousel_effect_card.dart）
///   - chips      标签胶囊墙（chips_effect_card.dart）
///   - fold_list  长清单折叠（fold_list_effect_card.dart）
///   - compare    A/B 双图拖动对比滑杆（compare_slider.dart）
/// 其余 cardType（weather/schedule/.../timeline/progress/quote/media/
/// search_result）仍由 agent_result_card.dart 内的既有组件渲染；
/// [displayEffectsCard] 返回 null 即表示「无专用效果，走原有默认」。
library;

import "package:flutter/material.dart";

import "../../../core/utils/agent_result_parser.dart";
import "carousel_effect_card.dart";
import "chips_effect_card.dart";
import "compare_slider.dart";
import "effect_media_utils.dart";
import "fold_list_effect_card.dart";
import "metric_effect_card.dart";
import "steps_effect_card.dart";

/// 按 [AgentResultData.cardType] 分发展示效果组件。
///
/// 返回 null 表示该 cardType 不在本模块管辖范围（调用方回退到
/// AgentResultCard 既有的专用卡/通用列表卡），保证新增类型前的
/// 历史行为完全不变。
Widget? displayEffectsCard({
  required AgentResultData data,
  required ColorScheme cs,
}) {
  switch (data.cardType) {
    case "steps":
      return StepsEffectCard(data: data, cs: cs);
    case "metric":
      return MetricEffectCard(data: data, cs: cs);
    case "carousel":
      return CarouselEffectCard(data: data, cs: cs);
    case "chips":
      return ChipsEffectCard(data: data, cs: cs);
    case "fold_list":
      return FoldListEffectCard(data: data, cs: cs);
    case "compare":
      return _buildCompare(data, cs);
    default:
      return null;
  }
}

/// compare：side=A / side=B 各一张可解析图片时启用双图滑杆；
/// 无 side 标注但恰好两条带图条目时按顺序视为 A/B；
/// 其余（纯文本对比、多图对比）返回 null 回退通用卡。
Widget? _buildCompare(AgentResultData data, ColorScheme cs) {
  String? urlA;
  String? urlB;
  for (final AgentResultItem it in data.items) {
    final String side = (it.side ?? "").trim().toUpperCase();
    if (side == "A" && urlA == null) {
      urlA = resolveItemPreviewUrl(it);
    } else if (side == "B" && urlB == null) {
      urlB = resolveItemPreviewUrl(it);
    }
  }
  if (urlA == null && urlB == null) {
    final List<String> urls = data.items
        .map(resolveItemPreviewUrl)
        .whereType<String>()
        .toList();
    if (urls.length == 2) {
      urlA = urls[0];
      urlB = urls[1];
    }
  }
  if (urlA == null || urlB == null) return null;
  return CompareEffectCard(data: data, cs: cs, urlA: urlA, urlB: urlB);
}

File: packages/agent-protocol/src/display-effects.ts
/**
 * 展示效果（display effects）契约。
 *
 * 服务端展示路由（server/src/services/display-effect-router.ts）按「工具
 * 信号 + 内容信号」从下列类型中选出 cardType，随 [AGENT_RESULT_CARD] 载荷
 * 下发；客户端（Flutter display_effects/ 模块 + agent_result_card.dart）
 * 按同名字符串分发到对应效果组件。本文件是两端共享的唯一枚举源——
 * 新增效果先在此登记，再同步路由器与前端组件，防止「服务端新增了、
 * 前端静默落通用卡」的漂移。
 */

/** 全部展示效果类型；空串 = 通用列表卡（前端默认）。 */
export type DisplayEffectType =
  | "weather" // 天气卡（工具：weather.*）
  | "schedule" // 日程卡（工具：calendar/schedule）
  | "wallet" // 钱包卡（工具：wallet.*）
  | "order" // 订单卡（工具：order/payment/alipay）
  | "file" // 文件卡（工具：file）
  | "search_result" // 搜索结果卡（工具：search_web/info.*）
  | "media" // 媒体图廊卡（工具：search_images/search_videos 等）
  | "compare" // A/B 双图对比滑杆（工具：compare/pk；或内容含对比词 + 图片）
  | "comparison_table" // 文本 A/B 对比双栏卡（内容：A/B 标签条目成对出现）
  | "timeline" // 时间轴卡（工具：plan/timeline；或条目以时间开头）
  | "progress" // 文字进度条卡（内容：百分比/分数占多数）
  | "steps" // 数字步骤卡（内容：第X步/Step N/数字. 开头占多数）
  | "metric" // 数据面板卡（内容：全部为「标签：数值」）
  | "carousel" // 轮播横滑卡（内容：多数条目内嵌图片 URL）
  | "chips" // 标签/徽章行（内容：全部为短标签）
  | "fold_list" // 折叠列表卡（内容：≥8 条长清单）
  | "quote" // 引用强调卡（markdown 引用块 / 引用式单句结论）
  | "travel_itinerary" // 旅游行程双面板卡（工具：travel.*）
  | "";

/** 展示效果类型的运行时清单（与 DisplayEffectType 逐一对应，供漂移测试）。 */
export const DISPLAY_EFFECT_TYPES = [
  "weather",
  "schedule",
  "wallet",
  "order",
  "file",
  "search_result",
  "media",
  "compare",
  "comparison_table",
  "timeline",
  "progress",
  "steps",
  "metric",
  "carousel",
  "chips",
  "fold_list",
  "quote",
  "travel_itinerary",
  "",
] as const satisfies readonly DisplayEffectType[];
